File: libs/shape/src/main/kotlin/mono/shape/connector/ShapeConnector.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.connector

import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.Point
import mono.graphics.geo.PointF
import mono.graphics.geo.Rect
import mono.shape.collection.TwoWayQuickMap
import mono.shape.connector.LineConnector.ConnectorIdentifier
import mono.shape.serialization.SerializableLineConnector
import mono.shape.shape.AbstractShape
import mono.shape.shape.Group
import mono.shape.shape.Line
import mono.shape.shape.MockShape
import mono.shape.shape.Rectangle
import mono.shape.shape.Text

/**
 * A manager of shape connectors.
 */
class ShapeConnector {
    private val lineConnectors = TwoWayQuickMap<LineConnector, AbstractShape.ShapeIdentifier>()

    fun addConnector(line: Line, anchor: Line.Anchor, shape: AbstractShape) {
        val anchorPoint = when (anchor) {
            Line.Anchor.START -> line.startPoint
            Line.Anchor.END -> line.endPoint
        }
        val boxBound = shape.bound
        val (ratio, offset) = calculateConnectorRatioAndOffset(anchorPoint, boxBound) ?: return

        val connector = LineConnector(line.id, anchor, ratio, offset)

        // TODO: Based on the position of anchor point and shape's bound, the direction of the
        //  anchor must be updated to optimise the initial directions of the line

        lineConnectors[connector] = shape.identifier
    }

    fun getConnectors(shape: AbstractShape): Collection<LineConnector> =
        lineConnectors.getKeys(shape)

    fun removeConnector(line: Line, anchor: Line.Anchor) =
        lineConnectors.removeKey(ConnectorIdentifier(line.id, anchor))

    fun hasConnector(line: Line, anchor: Line.Anchor): Boolean =
        lineConnectors[ConnectorIdentifier(line.id, anchor)] != null

    fun removeShape(shape: AbstractShape) {
        lineConnectors.removeValue(shape)

        when (shape) {
            is Line -> {
                removeConnector(shape, Line.Anchor.START)
                removeConnector(shape, Line.Anchor.END)
            }

            is Group,
            is MockShape,
            is Rectangle,
            is Text -> Unit
        }
    }

    /**
     * Calculates the connector ratio. Returns null if the two cannot connect.
     */
    private fun calculateConnectorRatioAndOffset(
        anchorPoint: DirectedPoint,
        boxBound: Rect
    ): Pair<PointF, Point>? {
        val around = ShapeConnectorUseCase.getAround(anchorPoint, boxBound) ?: return null

        val ratio = ShapeConnectorUseCase.calculateRatio(around, anchorPoint, boxBound)
        val offset = ShapeConnectorUseCase.calculateOffset(around, anchorPoint, boxBound)
        return ratio to offset
    }

    fun toSerializable(): List<SerializableLineConnector> =
        lineConnectors.asSequence()
            .map { (lineConnector, target) -> toSerializableConnector(lineConnector, target) }
            .toList()

    companion object {
        fun fromSerializable(
            serializedConnectors: List<SerializableLineConnector>
        ): ShapeConnector {
            val manager = ShapeConnector()
            for (conn in serializedConnectors) {
                val lineConnector = LineConnector(conn.lineId, conn.anchor, conn.ratio, conn.offset)
                manager.lineConnectors[lineConnector] = AbstractShape.ShapeIdentifier(conn.targetId)
            }
            return manager
        }

        fun toSerializableConnector(
            lineConnector: LineConnector,
            target: AbstractShape.ShapeIdentifier
        ): SerializableLineConnector = SerializableLineConnector(
            lineId = lineConnector.lineId,
            anchor = lineConnector.anchor,
            targetId = target.id,
            lineConnector.ratio,
            lineConnector.offset
        )

        fun toSerializableConnector(
            lineConnector: LineConnector,
            target: AbstractShape
        ): SerializableLineConnector = SerializableLineConnector(
            lineId = lineConnector.lineId,
            anchor = lineConnector.anchor,
            targetId = target.id,
            lineConnector.ratio,
            lineConnector.offset
        )
    }
}

File: monosketch-svelte/src/lib/mono/shape/connector/shape-connector.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { type DirectedPoint, Point, PointF } from "$libs/graphics-geo/point";
import type { Rect } from "$libs/graphics-geo/rect";
import { TwoWayQuickMap } from "$mono/shape/collection/two-way-quick-map";
import { ConnectorIdentifier, LineConnector, ShapeConnectorUseCase } from "$mono/shape/connector/line-connector";
import { SerializableLineConnector } from "$mono/shape/serialization/serializable-connector";
import { AbstractShape, ShapeIdentifier } from "$mono/shape/shape/abstract-shape";
import { Line } from "$mono/shape/shape/line";
import { LineAnchor } from "$mono/shape/shape/linehelper";

/**
 * A manager of shape connectors.
 */
export class ShapeConnector {
    private lineConnectors = new TwoWayQuickMap<LineConnector, ShapeIdentifier>();

    addConnector(line: Line, anchor: LineAnchor, shape: AbstractShape) {
        const anchorPoint = anchor === LineAnchor.START ? line.startPoint : line.endPoint;
        const boxBound = shape.bound;
        const result = this.calculateConnectorRatioAndOffset(anchorPoint, boxBound);
        if (!result) return;

        const [ratio, offset] = result;
        const connector = new LineConnector(line.id, anchor, ratio, offset);

        // TODO: Based on the position of anchor point and shape's bound, the direction of the
        //  anchor must be updated to optimise the initial directions of the line

        this.lineConnectors.set(connector, shape.identifier);
    }

    getConnectors(shape: AbstractShape): LineConnector[] {
        return Array.from(this.lineConnectors.getKeys(shape));
    }

    removeConnector(line: Line, anchor: LineAnchor) {
        this.lineConnectors.removeKey(new ConnectorIdentifier(line.id, anchor));
    }

    hasConnector(line: Line, anchor: LineAnchor): boolean {
        return this.lineConnectors.get(new ConnectorIdentifier(line.id, anchor)) !== undefined;
    }

    removeShape(shape: AbstractShape) {
        this.lineConnectors.removeValue(shape);

        if (shape instanceof Line) {
            this.removeConnector(shape, LineAnchor.START);
            this.removeConnector(shape, LineAnchor.END);
        }
    }

    /**
     * Calculates the connector ratio. Returns null if the two cannot connect.
     */
    private calculateConnectorRatioAndOffset(anchorPoint: DirectedPoint, boxBound: Rect): [PointF, Point] | null {
        const around = ShapeConnectorUseCase.getAround(anchorPoint, boxBound);
        if (!around) {
            return null
        }

        const ratio = ShapeConnectorUseCase.calculateRatio(around, anchorPoint, boxBound);
        const offset = ShapeConnectorUseCase.calculateOffset(around, anchorPoint, boxBound);
        return [ratio, offset];
    }

    toSerializable(): SerializableLineConnector[] {
        return Array.from(this.lineConnectors.asSequence()).map(([lineConnector, target]) =>
            ShapeConnector.toSerializableConnector(lineConnector, target),
        );
    }

    static fromSerializable(serializedConnectors: SerializableLineConnector[]): ShapeConnector {
        const manager = new ShapeConnector();
        for (const conn of serializedConnectors) {
            const lineConnector = new LineConnector(conn.lineId, conn.anchor, conn.ratio, conn.offset);
            manager.lineConnectors.set(lineConnector, new ShapeIdentifier(conn.targetId));
        }
        return manager;
    }

    static toSerializableConnector(
        lineConnector: LineConnector,
        targetOrIdentifier: ShapeIdentifier | AbstractShape,
    ): SerializableLineConnector {
        if (targetOrIdentifier instanceof AbstractShape) {
            return new SerializableLineConnector(
                lineConnector.lineId,
                lineConnector.anchor,
                targetOrIdentifier.id,
                lineConnector.ratio,
                lineConnector.offset,
            );
        } else {
            return new SerializableLineConnector(
                lineConnector.lineId,
                lineConnector.anchor,
                targetOrIdentifier.id,
                lineConnector.ratio,
                lineConnector.offset,
            );
        }
    }
}
